### PR TITLE
TINY-8139: Replace parser

### PR DESCRIPTION
**Pull request by ashfordneil**
Related Ticket: TINY-8139

Description of Changes:
* Turned DomParser (tinymce) into a wrapper around DOMParser (native)
* Commented out or deleted anything that broke
* Updated the insert content and get content methods to append DOM elements manually, instead of re-serializing them

Pre-checks:
* [x] ~Changelog entry added~
* [x] ~Tests have been added (if applicable)~ The tests won't even run at this point
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable): N/A

### CODE REVIEW

File: modules/tinymce/src/core/main/ts/dom/DomSerializerImpl.ts [hunk 1]
```diff
@@ -106,7 +107,13 @@ const DomSerializerImpl = (settings: DomSerializerSettings, editor: Editor): Dom
     const targetNode = DomSerializerPreProcess.process(editor, node, args);
     const html = getHtmlFromNode(dom, targetNode, args);
     const rootNode = parseHtml(htmlParser, html, args);
-    return args.format === 'tree' ? rootNode : toHtml(editor, settings, schema, rootNode, args);
+    if (args.format === 'tree') {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore need to figure out how to fake the tree format
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/dom/DomSerializerImpl.ts:112]: Can you please make sure to add TODOs with the epic Jira number in places you've commented out or done this to make sure we don't forget/miss any?

- [ ] Add TODOs with epic Jira numbers for commented out/asserted code

File: modules/tinymce/src/core/test/ts/browser/content/InsertListTest.ts [hunk 2]
```diff
@@ -1,48 +0,0 @@
-import { describe, it } from '@ephox/bedrock-client';
-import { assert } from 'chai';
-
-import DOMUtils from 'tinymce/core/api/dom/DOMUtils';
-import DomParser from 'tinymce/core/api/html/DomParser';
-import AstNode from 'tinymce/core/api/html/Node';
-import Schema from 'tinymce/core/api/html/Schema';
-import * as InsertList from 'tinymce/core/content/InsertList';
-
-describe('browser.tinymce.core.content.InsertListTest', () => {
```
Comment (lnewson) [inline modules/tinymce/src/core/test/ts/browser/content/InsertListTest.ts:10]: Why was this deleted? Based on what it's doing it shouldn't have been removed.

- [ ] Revert or provide a good reason as to why this was deleted

File: modules/tinymce/src/core/main/ts/api/html/DomParser.ts [hunk 3]
```diff
@@ -80,154 +78,19 @@ interface DomParser {
   addNodeFilter: (name: string, callback: (nodes: AstNode[], name: string, args: ParserArgs) => void) => void;
   getNodeFilters: () => ParserFilter[];
   filterNode: (node: AstNode) => AstNode;
-  parse: (html: string, args?: ParserArgs) => AstNode;
+  parse: (html: string, args?: ParserArgs) => Node;
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/html/DomParser.ts:81]: This needs to return an `AstNode` still, so please make sure to either fix that or leave a TODO for when we have the wrapper.

File: modules/tinymce/src/core/main/ts/api/html/DomParser.ts [hunk 4]
```diff
@@ -80,154 +78,19 @@ interface DomParser {
...
-  let matchedNodes: Record<string, AstNode[]> = {};
-  let matchedAttributes: Record<string, AstNode[]> = {};
+  const matchedNodes: Record<string, AstNode[]> = {};
+  const matchedAttributes: Record<string, AstNode[]> = {};
 
   settings = settings || {};
   settings.validate = 'validate' in settings ? settings.validate : true;
   settings.root_name = settings.root_name || 'body';
 
-  const fixInvalidChildren = (nodes: AstNode[]) => {
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/html/DomParser.ts:96]: All of this logic shouldn't have been deleted, as we're going to need to retain it (or similar logic) long term.

- [ ] Revert this and either make it work or comment it out

File: modules/tinymce/src/core/main/ts/api/html/DomParser.ts [hunk 5]
```diff
@@ -343,422 +206,15 @@ const DomParser = (settings?: DomParserSettings, schema = Schema()): DomParser =
...
-          }
-
-          for (fi = 0, fl = list.callbacks.length; fi < fl; fi++) {
-            list.callbacks[fi](nodes, list.name, args);
-          }
-        }
-      }
-    }
+  const parse = (html: string, _args?: ParserArgs): Node => {
+    const parser = new DOMParser();
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/html/DomParser.ts:213]: This doesn't need to be constructed each time and could be a global instead. That'll help with parse performance.

- [ ] Make this a global variable

File: modules/tinymce/src/core/main/ts/api/html/DomParser.ts [hunk 6]
```diff
@@ -343,422 +206,15 @@ const DomParser = (settings?: DomParserSettings, schema = Schema()): DomParser =
...
+    const doc = parser.parseFromString(html, 'text/html');
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/html/DomParser.ts:214]: This doesn't account for the `element_format` which means we might need to parse as XHTML. Additionally, it doesn't account for things like the `context` passed in via the `args` so this needs a lot more work sorry.

- [ ] Ensure the `DomParserSettings` and `ParserArgs` are used appropriately

**Review by lnewson — CHANGES_REQUESTED** [review]

**Comment by lnewson** [comment]
There are some things missing but is a good start Neil :+1:

File: modules/tinymce/src/core/test/ts/browser/html/DomParserTest.ts [hunk 7]
```diff
@@ -1,819 +0,0 @@
...
-import { assert } from 'chai';
-
-import Env from 'tinymce/core/api/Env';
-import { BlobCache } from 'tinymce/core/api/file/BlobCache';
-import DomParser from 'tinymce/core/api/html/DomParser';
-import AstNode from 'tinymce/core/api/html/Node';
-import Schema from 'tinymce/core/api/html/Schema';
-import HtmlSerializer from 'tinymce/core/api/html/Serializer';
-
-describe('browser.tinymce.core.html.DomParserTest', () => {
```
Comment (lnewson) [inline modules/tinymce/src/core/test/ts/browser/html/DomParserTest.ts:12]: This test shouldn't have been deleted, or at least not entirely as some of it applies to validation and "fixing" nodes. Please revert this and either skip it or comment it out instead so we can determine what should be kept.

- [ ] Revert deleting this test and skip/comment instead

File: modules/tinymce/src/core/main/ts/api/html/DomParser.ts [hunk 8]
```diff
@@ -343,422 +206,15 @@ const DomParser = (settings?: DomParserSettings, schema = Schema()): DomParser =
...
-      for (const name in input) {
-        if (name !== 'li' && name !== 'p') {
-          output[name] = input[name];
-        }
-      }
-
-      return output;
-    };
-
-    const parser = SaxParser({
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/html/DomParser.ts:493]: How come `SaxParser` wasn't removed as part of this PR?

File: modules/tinymce/src/core/main/ts/dom/DomSerializerImpl.ts (hunk 1 above)
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/dom/DomSerializerImpl.ts:115]: @ashfordneil this hasn't been done, as I asked for it to contain the **Epic Jira** numbers, so we can search for a single thing to make sure all the work was done. In addition to that a number of TODO statements have an invalid JIRA number (TINY-8902). If you want the others, then please include both JIRA numbers but I definitely want the Epic in there.

File: modules/tinymce/src/core/main/ts/api/html/DomParser.ts (hunk 3 above)
Comment (ashfordneil) [reply modules/tinymce/src/core/main/ts/api/html/DomParser.ts:84]: Todo added (now with new and improved _correct_ Jira numbers)

File: modules/tinymce/src/core/main/ts/dom/DomSerializerImpl.ts (hunk 1 above)
Comment (ashfordneil) [reply modules/tinymce/src/core/main/ts/dom/DomSerializerImpl.ts:115]: Should be good now. Thanks for the catch on the Typos that I'd made with invalid Jira numbers

File: modules/tinymce/src/core/test/ts/browser/content/InsertListTest.ts (hunk 2 above)
Comment (ashfordneil) [reply modules/tinymce/src/core/test/ts/browser/content/InsertListTest.ts:10]: It's back (just skipped with some more `any` types and TODOs)

File: modules/tinymce/src/core/test/ts/browser/html/DomParserTest.ts (hunk 7 above)
Comment (ashfordneil) [reply modules/tinymce/src/core/test/ts/browser/html/DomParserTest.ts:12]: Back, with Jiras to fix it up

File: modules/tinymce/src/core/main/ts/api/html/DomParser.ts (hunk 4 above)
Comment (ashfordneil) [reply modules/tinymce/src/core/main/ts/api/html/DomParser.ts]: It's back, albeit unused right now, with comments saying to fold this logic into the new system.

File: modules/tinymce/src/core/main/ts/api/html/DomParser.ts (hunk 5 above)
Comment (ashfordneil) [reply modules/tinymce/src/core/main/ts/api/html/DomParser.ts]: It's now a (lazily instantiated) global

File: modules/tinymce/src/core/main/ts/api/html/DomParser.ts (hunk 6 above)
Comment (ashfordneil) [reply modules/tinymce/src/core/main/ts/api/html/DomParser.ts]: TODO added

File: modules/tinymce/src/core/main/ts/api/html/DomParser.ts (hunk 8 above)
Comment (ashfordneil) [reply modules/tinymce/src/core/main/ts/api/html/DomParser.ts]: I've got a task inside this epic to delete the SaxParser.

Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/html/DomParser.ts]: Why not just add the TODO to SaxParser as I really want a TODO in there at the very least?

File: modules/tinymce/src/core/main/ts/api/html/DomParser.ts [hunk 9]
```diff
@@ -80,9 +80,12 @@ interface DomParser {
   addNodeFilter: (name: string, callback: (nodes: AstNode[], name: string, args: ParserArgs) => void) => void;
   getNodeFilters: () => ParserFilter[];
   filterNode: (node: AstNode) => AstNode;
-  parse: (html: string, args?: ParserArgs) => AstNode;
+  // TODO: TINY-4627 (TINY-8190) make this back into an AstNode, but in a good way
+  parse: (html: string, args?: ParserArgs) => Node;
 }
 
+let lazyParser: DOMParser | null = null;
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/html/DomParser.ts:87]: Can you please change this to a `Thunk.cached` call instead? That's our normal way of doing something that has to lazily initialise and only run once.

File: modules/tinymce/src/core/main/ts/api/html/DomParser.ts (hunk 5 above)
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/html/DomParser.ts]: I've left a separate comment about how the lazy instantiation was done so resolving this comment.

**Review by lnewson — APPROVED** [review]

**Review by metricjs — APPROVED** [review]

**Comment by ashfordneil** [comment]
Closing this in favour of a more gradual approach, starting with #7448 and continuing on with some more PRs
